third-party/fb-mysql/8.0.20/man/mysql_ssl_rsa_setup.1
'\" t
.\"     Title: \fBmysql_ssl_rsa_setup\fR
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 03/06/2020
.\"    Manual: MySQL Database System
.\"    Source: MySQL 8.0
.\"  Language: English
.\"
.TH "\FBMYSQL_SSL_RSA_SETUP\FR" "1" "03/06/2020" "MySQL 8\&.0" "MySQL Database System"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
mysql_ssl_rsa_setup \- create SSL/RSA files
.SH "SYNOPSIS"
.HP \w'\fBmysql_ssl_rsa_setup\ [\fR\fB\fIoptions\fR\fR\fB]\fR\ 'u
\fBmysql_ssl_rsa_setup [\fR\fB\fIoptions\fR\fR\fB]\fR
.SH "DESCRIPTION"
.PP
This program creates the SSL certificate and key files and RSA key\-pair files required to support secure connections using SSL and secure password exchange using RSA over unencrypted connections, if those files are missing\&.
\fBmysql_ssl_rsa_setup\fR
can also be used to create new SSL files if the existing ones have expired\&.
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
.PP
\fBmysql_ssl_rsa_setup\fR
uses the
\fBopenssl\fR
command, so its use is contingent on having OpenSSL installed on your machine\&.
.PP
Another way to generate SSL and RSA files, for MySQL distributions compiled using OpenSSL, is to have the server generate them automatically\&. See
Section\ \&6.3.3.1, \(lqCreating SSL and RSA Certificates and Keys using MySQL\(rq\&.
.sp .5v
.RE
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBImportant\fR
.ps -1
.br
.PP
\fBmysql_ssl_rsa_setup\fR
helps lower the barrier to using SSL by making it easier to generate the required files\&. However, certificates generated by
\fBmysql_ssl_rsa_setup\fR
are self\-signed, which is not very secure\&. After you gain experience using the files created by
\fBmysql_ssl_rsa_setup\fR, consider obtaining a CA certificate from a registered certificate authority\&.
.sp .5v
.RE
.PP
Invoke
\fBmysql_ssl_rsa_setup\fR
like this:
.sp
.if n \{\
.RS 4
.\}
.nf
shell> \fBmysql_ssl_rsa_setup [\fR\fB\fIoptions\fR\fR\fB]\fR
.fi
.if n \{\
.RE
.\}
.PP
Typical options are
\fB\-\-datadir\fR
to specify where to create the files, and
\fB\-\-verbose\fR
to see the
\fBopenssl\fR
commands that
\fBmysql_ssl_rsa_setup\fR
executes\&.
.PP
\fBmysql_ssl_rsa_setup\fR
attempts to create SSL and RSA files using a default set of file names\&. It works as follows:
.sp
.RS 4
.ie n \{\
\h'-04' 1.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  1." 4.2
.\}
\fBmysql_ssl_rsa_setup\fR
checks for the
\fBopenssl\fR
binary at the locations specified by the
PATH
environment variable\&. If
\fBopenssl\fR
is not found,
\fBmysql_ssl_rsa_setup\fR
does nothing\&. If
\fBopenssl\fR
is present,
\fBmysql_ssl_rsa_setup\fR
looks for default SSL and RSA files in the MySQL data directory specified by the
\fB\-\-datadir\fR
option, or the compiled\-in data directory if the
\fB\-\-datadir\fR
option is not given\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04' 2.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  2." 4.2
.\}
\fBmysql_ssl_rsa_setup\fR
checks the data directory for SSL files with the following names:
.sp
.if n \{\
.RS 4
.\}
.nf
ca\&.pem
server\-cert\&.pem
server\-key\&.pem
.fi
.if n \{\
.RE
.\}
.RE
.sp
.RS 4
.ie n \{\
\h'-04' 3.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  3." 4.2
.\}
If any of those files are present,
\fBmysql_ssl_rsa_setup\fR
creates no SSL files\&. Otherwise, it invokes
\fBopenssl\fR
to create them, plus some additional files:
.sp
.if n \{\
.RS 4
.\}
.nf
ca\&.pem               Self\-signed CA certificate
ca\-key\&.pem           CA private key
server\-cert\&.pem      Server certificate
server\-key\&.pem       Server private key
client\-cert\&.pem      Client certificate
client\-key\&.pem       Client private key
.fi
.if n \{\
.RE
.\}
.sp
These files enable secure client connections using SSL; see
Section\ \&6.3.1, \(lqConfiguring MySQL to Use Encrypted Connections\(rq\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04' 4.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  4." 4.2
.\}
\fBmysql_ssl_rsa_setup\fR
checks the data directory for RSA files with the following names:
.sp
.if n \{\
.RS 4
.\}
.nf
private_key\&.pem      Private member of private/public key pair
public_key\&.pem       Public member of private/public key pair
.fi
.if n \{\
.RE
.\}
.RE
.sp
.RS 4
.ie n \{\
\h'-04' 5.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  5." 4.2
.\}
If any of these files are present,
\fBmysql_ssl_rsa_setup\fR
creates no RSA files\&. Otherwise, it invokes
\fBopenssl\fR
to create them\&. These files enable secure password exchange using RSA over unencrypted connections for accounts authenticated by the
sha256_password
or
caching_sha2_password
plugin; see
Section\ \&6.4.1.3, \(lqSHA-256 Pluggable Authentication\(rq, and
Section\ \&6.4.1.2, \(lqCaching SHA-2 Pluggable Authentication\(rq\&.
.RE
.PP
For information about the characteristics of files created by
\fBmysql_ssl_rsa_setup\fR, see
Section\ \&6.3.3.1, \(lqCreating SSL and RSA Certificates and Keys using MySQL\(rq\&.
.PP
At startup, the MySQL server automatically uses the SSL files created by
\fBmysql_ssl_rsa_setup\fR
to enable SSL if no explicit SSL options are given other than
\fB\-\-ssl\fR
(possibly along with
\fB\-\-ssl\-cipher\fR)\&. If you prefer to designate the files explicitly, invoke clients with the
\fB\-\-ssl\-ca\fR,
\fB\-\-ssl\-cert\fR, and
\fB\-\-ssl\-key\fR
options at startup to name the
ca\&.pem,
server\-cert\&.pem, and
server\-key\&.pem
files, respectively\&.
.PP
The server also automatically uses the RSA files created by
\fBmysql_ssl_rsa_setup\fR
to enable RSA if no explicit RSA options are given\&.
.PP
If the server is SSL\-enabled, clients use SSL by default for the connection\&. To specify certificate and key files explicitly, use the
\fB\-\-ssl\-ca\fR,
\fB\-\-ssl\-cert\fR, and
\fB\-\-ssl\-key\fR
options to name the
ca\&.pem,
client\-cert\&.pem, and
client\-key\&.pem
files, respectively\&. However, some additional client setup may be required first because
\fBmysql_ssl_rsa_setup\fR
by default creates those files in the data directory\&. The permissions for the data directory normally enable access only to the system account that runs the MySQL server, so client programs cannot use files located there\&. To make the files available, copy them to a directory that is readable (but
\fInot\fR
writable) by clients:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}

NEWS
"Rogers" 10-Sep-2015
- Update to jemalloc 4.0.0.
- Convert more architecture dependent code to VASM.
- Large bout of fixes for compiling hhvm under MSVC.
- Have hh_emitter support more of HHBC.
- Remove MInstrState from CGetM and IssetM's final operations.
- Refactor [un]serialize functions.
- Use profiling counters to decide whether to apply LICM.
- Move inlining to translateRegion.
- Implement intdiv function.
- Remove joomla and joomlacms from hhvm test suite.
- [hack] Add a --stats option to the client.
- [hack] GC the shared memory area more aggressively.
- [hack] Rework the representation of strings.
- Fix bug #70014 - use RAND_bytes instead of deprecated RAND_pseudo_bytes
- Fix bug #69545 - avoid overflow when reading list and improve fix for bug #69545
- Fix #66387: Stack overflow with imagefilltoborder
- Fix #70064: imagescale(..., IMG_BICUBIC) leaks memory and Fix #69024: imagescale segfault with palette based image

"Quake" 27-Aug-2015
- ??

"Park" 13-Aug-2015
- ???

"Osborne" 30-Jul-2015
- Remove global LDAP requirement
- Don't require libXSLT globally
- Remove the global MCrypt requirement
- Remove the global search for LibUODBC
- Remove the global check for libpam
- use folly::writeFull wrapper for uploads
- Fix vararg handling for array_uintersect and others
- fix error message for abstract type constants
- Fix variadics in hhvm_get_frame_args
- Use poll() instead of select() in XDebugServer::createSocket.
- [hack] Error on overriding concrete type constant with an abstract one
- [hack] Grow shapes more correctly
- Array pointer modification functions (next, current, each etc) can be used on ArrayObject
- Eliminate ext_mbstring's dependence on ext_process
- Eliminate ext_mail's dependence on ext_process
- Use utf8 string literals in zend-html
- Don't allow abstract memoized methods
- [hack] Properly handle 'this' type for methods
- Fix lexing of "${fn()}"
- Fix lexing of #<?php
- [hack] Fix infinite loop in syntax error when closing a namespace
- Fix mispelling of COOKIE.
- Deal with self, static and parent constructors

"Nash" 16-Jul-2015
- Make ArrayData ref-count start at 1
- Remove the NOUSER define when compiling with MSVC, Cygwin, or MinGW
- Fix typos of `__MSC_VER` to be the correct value, `_MSC_VER`
- Shapes::removeKey(...) - remember the names of unset fields
- Shapes::toArray() - add .hhi
- Shapes::removeKey()
- Shapes::keyExists()
- Shapes::idx helper function for accessing optional fields
- Don't assume a scalar Map/Set initializer is an integer if its not a string
- Update mysqlclient library to 5.6.23 everywhere
- Port fix for php bug #48147
- Fixed #5475: ResourceBundle::getLocales('') don't return list of locales
- [hhvm][oss][frameworks] Add Smarty to Framework test runner
- Add in a facility to pool curl handles between requests.
- Fix xdebug trying to write to a nullptr std::FILE*.
- Fix bug when appending unaligned StringSlice
- Remove XDebug chrome support
- Incorrect return value from eval, Closure generated in first eval pass is returned in the second eval pass
- Stop throwing ints in xdebug
- CVE-2015-4663: Support verifying certificates on ssl/tls streams, enable by default
- Allow inlining immutable methods in non-repo mode
- Reusable TC
- Release some old hack design documentation
- Don't step into builtin functions in the debugger
- Support collections in array_map.hhas
- [hhvm] Make sure iteration of class_iterator actually stops
- [hhvm] Open source idx
- Add .hhas implementation of array_reduce
- Change memory layout for hni

"McQueen" 2-Jul-2015
- Add basic stream_socket_enable_crypto() client support
- [hack] Check types for class vars in decl mode as well

third-party/fb-mysql/8.0.20/client/mysql_ssl_rsa_setup.cc
/*
   Copyright (c) 2015, 2019, Oracle and/or its affiliates. All rights reserved.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License, version 2.0,
   as published by the Free Software Foundation.

   This program is also distributed with certain software (including
   but not limited to OpenSSL) that is licensed under separate terms,
   as designated in a particular file or component or in included license
   documentation.  The authors of MySQL hereby grant you an additional
   permission to link the program and your derivative works with the
   separately licensed software that they have included with MySQL.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License, version 2.0, for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA
*/

#include "my_config.h"

#include <mysql_version.h>
#include <stdint.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <algorithm>
#include <cstdio>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>

#include "client/logger.h"
#include "client/path.h"
#ifdef _WIN32
#include "m_ctype.h"
#endif
#include "my_alloc.h"
#include "my_compiler.h"
#include "my_dbug.h"
#include "my_default.h"
#include "my_dir.h"
#include "my_getopt.h"
#include "my_inttypes.h"
#include "my_io.h"
#include "my_macros.h"
#include "my_sys.h"
#include "mysql/service_mysql_alloc.h"
#include "print_version.h"
#include "welcome_copyright_notice.h" /* ORACLE_WELCOME_COPYRIGHT_NOTICE */

#if HAVE_CHOWN
#include <pwd.h>
#endif

/* Forward declarations */

using namespace std;
typedef string Sql_string_t;

static Sql_string_t create_string(const char *ptr);

/* Global Variables */
enum certs {
  CA_CERT = 0,
  CA_KEY,
  CA_REQ,
  SERVER_CERT,
  SERVER_KEY,
  SERVER_REQ,
  CLIENT_CERT,
  CLIENT_KEY,
  CLIENT_REQ,
  PRIVATE_KEY,
  PUBLIC_KEY,
  OPENSSL_RND
};

enum extfiles { CAV3_EXT = 0, CERTV3_EXT };

Sql_string_t cert_files[] = {
    create_string("ca.pem"),          create_string("ca-key.pem"),
    create_string("ca-req.pem"),      create_string("server-cert.pem"),
    create_string("server-key.pem"),  create_string("server-req.pem"),
    create_string("client-cert.pem"), create_string("client-key.pem"),
    create_string("client-req.pem"),  create_string("private_key.pem"),
    create_string("public_key.pem"),  create_string(".rnd")};

Sql_string_t ext_files[] = {create_string("cav3.ext"),
                            create_string("certv3.ext")};

#define MAX_PATH_LEN \
  (FN_REFLEN - strlen(FN_DIRSEP) - cert_files[SERVER_CERT].length() - 1)
/*
  Higest number of fixed characters in subject line is 47:
  MySQL_SERVER_<suffix>_Auto_Generated_Server_Certificate
  Maximum size of subject is 64. So suffix can't be longer
  than 17 characters.
*/
#define MAX_SUFFIX_LEN 17

Log info(cout, "NOTE");
Log error(cerr, "ERROR");

char **defaults_argv = nullptr;
static char *opt_datadir = nullptr;
static char default_data_dir[] = MYSQL_DATADIR;
static char *opt_suffix = nullptr;
static char default_suffix[] = MYSQL_SERVER_VERSION;
#if HAVE_CHOWN
static char *opt_userid = nullptr;
struct passwd *user_info = nullptr;
#endif /* HAVE_CHOWN */
Path dir_string;
Sql_string_t suffix_string;
bool opt_verbose;

static const char *load_default_groups[] = {"mysql_ssl_rsa_setup", "mysqld",
                                            nullptr};

static struct my_option my_options[] = {
    {"help", '?', "Display this help and exit.", nullptr, nullptr, nullptr,
     GET_NO_ARG, NO_ARG, 0, 0, 0, nullptr, 0, nullptr},
    {"verbose", 'v', "Be more verbose when running program", &opt_verbose,
     nullptr, nullptr, GET_BOOL, NO_ARG, false, 0, 0, nullptr, 0, nullptr},
    {"version", 'V', "Print program version and exit", nullptr, nullptr,
     nullptr, GET_NO_ARG, NO_ARG, 0, 0, 0, nullptr, 0, nullptr},
    {"datadir", 'd', "Directory to store generated files.", &opt_datadir,
     &opt_datadir, nullptr, GET_STR_ALLOC, REQUIRED_ARG,
     (longlong)&default_data_dir, 0, 0, nullptr, 0, nullptr},
    {"suffix", 's', "Suffix to be added in certificate subject line",
     &opt_suffix, &opt_suffix, nullptr, GET_STR_ALLOC, REQUIRED_ARG,
     (longlong)&default_suffix, 0, 0, nullptr, 0, nullptr},
#if HAVE_CHOWN
    {"uid", 0, "The effective user id to be used for file permission",
     &opt_userid, &opt_userid, nullptr, GET_STR_ALLOC, REQUIRED_ARG, 0, 0, 0,
     nullptr, 0, nullptr},
#endif /* HAVE_CHOWN */
    /* END TOKEN */
    {nullptr, 0, nullptr, nullptr, nullptr, nullptr, GET_NO_ARG, NO_ARG, 0, 0,
     0, nullptr, 0, nullptr}};

/* Helper Functions */

/**
  The string class will break if constructed with a NULL pointer. This wrapper
  provides a systematic protection when importing char pointers.
 */
static Sql_string_t create_string(const char *ptr) {
  return (ptr ? Sql_string_t(ptr) : Sql_string_t(""));
}

static int execute_command(const Sql_string_t &command,
                           const Sql_string_t &error_message) {
  stringstream cmd_string;

  cmd_string << command;
  if (!opt_verbose) {
#ifndef _WIN32
    cmd_string << " > /dev/null 2>&1";
#else
    cmd_string << " > NUL 2>&1";
#endif /* _WIN32 */
  }

  info << "Executing : " << cmd_string.str() << endl;
  if (system(cmd_string.str().c_str())) {
    error << error_message << endl;
    return 1;
  }

  return 0;
}

static int set_file_pair_permission(const Sql_string_t &priv,
                                    const Sql_string_t &pub) {
  if (my_chmod(priv.c_str(), USER_READ | USER_WRITE, MYF(MY_FAE + MY_WME)) ||
      my_chmod(pub.c_str(), USER_READ | USER_WRITE | GROUP_READ | OTHERS_READ,
               MYF(MY_FAE + MY_WME))) {
    error << "Error setting file permissions for" << priv.c_str() << " and "
          << pub.c_str() << endl;
    return 1;
  }
#if HAVE_CHOWN
  if (user_info) {
    if (chown(priv.c_str(), user_info->pw_uid, user_info->pw_gid) ||
        chown(pub.c_str(), user_info->pw_uid, user_info->pw_gid)) {
      error << "Failed to change file permission" << endl;
      return 1;
    }
  }
#endif /* HAVE_CHOWN */
  return 0;
}

static bool file_exists(const Sql_string_t &filename) {
  MY_STAT file_stat;
  if (my_stat(filename.c_str(), &file_stat, MYF(0)) == nullptr) return false;

  return true;
}

static int remove_file(const Sql_string_t &filename, bool report_error = true) {
  if (my_delete(filename.c_str(), MYF(0))) {
    if (report_error) error << "Error deleting : " << filename << endl;
    return 1;
  }
  return 0;
}

static void free_resources() {
  if (opt_datadir) my_free(opt_datadir);
  if (opt_suffix) my_free(opt_suffix);
#if HAVE_CHOWN
  if (opt_userid) my_free(opt_userid);
#endif
}

class RSA_priv {
 public:
  RSA_priv(uint32_t key_size = 2048) : m_key_size(key_size) {}

  ~RSA_priv() {}

  Sql_string_t operator()(const Sql_string_t &key_file) {
    stringstream command;
    command << "openssl genrsa "
            << " -out " << key_file << " " << m_key_size;

    return command.str();
  }

 private:
  uint32_t m_key_size;
};

class RSA_pub {
 public:
  Sql_string_t operator()(const Sql_string_t &priv_key_file,
                          const Sql_string_t &pub_key_file) {
    stringstream command;
    command << "openssl rsa -in " << priv_key_file << " -pubout -out "
            << pub_key_file;
    return command.str();
  }
};

class X509_key {
 public:
  X509_key(const Sql_string_t &version, uint32_t validity = 10 * 365L)
      : m_validity(validity) {
    m_subj_prefix << "-subj /CN=MySQL_Server_" << version;
  }

  Sql_string_t operator()(Sql_string_t suffix, const Sql_string_t &key_file,
                          const Sql_string_t &req_file) {
    stringstream command;
    command << "openssl req -newkey rsa:2048 -days " << m_validity
            << " -nodes -keyout " << key_file << " " << m_subj_prefix.str()
            << suffix << " -out " << req_file << " && openssl rsa -in "
            << key_file << " -out " << key_file;

    return command.str();
  }

 private:
  uint32_t m_validity;
  stringstream m_subj_prefix;
};

class X509v3_ext_writer {
 public:
  X509v3_ext_writer() {
    m_cav3_ext_options << "basicConstraints=CA:TRUE" << std::endl;

    m_certv3_ext_options << "basicConstraints=CA:FALSE" << std::endl;
  }
  ~X509v3_ext_writer() {}

  bool operator()(const Sql_string_t &cav3_ext_file,
                  const Sql_string_t &certv3_ext_file) {
    if (!cav3_ext_file.length() || !certv3_ext_file.length()) return true;

    std::ofstream ext_file;

    ext_file.open(cav3_ext_file.c_str(), std::ios::out | std::ios::trunc);
    if (!ext_file.is_open()) return true;
    ext_file << m_cav3_ext_options.str();
    ext_file.close();

    ext_file.open(certv3_ext_file.c_str(), std::ios::out | std::ios::trunc);
    if (!ext_file.is_open()) {
      remove_file(cav3_ext_file.c_str(), false);
      return true;
    }
    ext_file << m_certv3_ext_options.str();
    ext_file.close();

    return false;
  }

 private:
  stringstream m_cav3_ext_options;
  stringstream m_certv3_ext_options;
};

class X509_cert {
 public:
  X509_cert(uint32_t validity = 10 * 365L) : m_validity(validity) {}

  ~X509_cert() {}

  Sql_string_t operator()(const Sql_string_t &req_file,
                          const Sql_string_t &cert_file, uint32_t serial,
                          bool self_signed, const Sql_string_t &sign_key_file,
                          const Sql_string_t &sign_cert_file,
                          const Sql_string_t &ext_file) {
    stringstream command;
    command << "openssl x509 -sha256 -days " << m_validity;
    command << " -extfile " << ext_file;
    command << " -set_serial " << serial << " -req -in " << req_file;
    if (self_signed)
      command << " -signkey " << sign_key_file;
    else
      command << " -CA " << sign_cert_file << " -CAkey " << sign_key_file;
    command << " -out " << cert_file;

    return command.str();
  }

 protected:
  uint32_t m_validity;
};

static void usage(void) {
  print_version();
  cout << (ORACLE_WELCOME_COPYRIGHT_NOTICE("2015")) << endl
       << "MySQL SSL Certificate and RSA Key Generation Utility" << endl
       << "Usage : " << my_progname << " [OPTIONS]" << endl;

  my_print_help(my_options);
  my_print_variables(my_options);
}

extern "C" {
static bool my_arguments_get_one_option(int optid, const struct my_option *,
                                        char *) {
  switch (optid) {
    case '?':
      usage();
      free_resources();
      exit(0);
    case 'V':
      print_version();
      free_resources();
      exit(0);
  }
  return false;
}
}

static inline bool is_not_alnum_underscore(char c) {
  return !(isalnum(c) || c == '_');
}

static bool check_suffix() {
  return (strcmp(opt_suffix, default_suffix) &&
          (find_if(suffix_string.begin(), suffix_string.end(),
                   is_not_alnum_underscore) != suffix_string.end()));
}

int main(int argc, char *argv[]) {
  int ret_val = 0;
  Sql_string_t openssl_check("openssl version");
  bool save_skip_unknown = my_getopt_skip_unknown;
  MEM_ROOT alloc{PSI_NOT_INSTRUMENTED, 512};

  MY_INIT(argv[0]);
  DBUG_TRACE;
  DBUG_PROCESS(argv[0]);

  /* Parse options : Command Line/Config file */

#ifdef _WIN32
  /* Convert command line parameters from UTF16LE to UTF8MB4. */
  my_win_translate_command_line_args(&my_charset_utf8mb4_bin, &argc, &argv);
#endif
  my_getopt_use_args_separator = true;
  if (load_defaults("my", load_default_groups, &argc, &argv, &alloc)) {
    my_end(0);
    free_resources();
    exit(1);
  }

  MY_MODE file_creation_mode = get_file_perm(USER_READ | USER_WRITE);
  MY_MODE saved_umask = umask(~(file_creation_mode));

  defaults_argv = argv;
  my_getopt_use_args_separator = false;
  my_getopt_skip_unknown = true;

  if (handle_options(&argc, &argv, my_options, my_arguments_get_one_option)) {
    error << "Error parsing options" << endl;
    ret_val = 1;
    goto end;
  }

  my_getopt_skip_unknown = save_skip_unknown;

  /* Process opt_verbose */
  if (opt_verbose != true) info.enabled(false);

  /* Process opt_datadir */

  dir_string.path(create_string(opt_datadir));
  if (dir_string.to_str().length() > MAX_PATH_LEN) {
    error << "Dir path is too long" << endl;
    ret_val = 1;
    goto end;
  }

  if (!dir_string.normalize_path() || !dir_string.exists()) {
    error << "Failed to access directory pointed by --datadir. "
          << "Please make sure that directory exists and is "
          << "accessible by mysql_ssl_rsa_setup. Supplied value : "
          << dir_string.to_str() << endl;
    ret_val = 1;
    goto end;
  }

  info << "Destination directory: " << dir_string.to_str() << endl;

  /* Process opt_suffix */

  suffix_string.append(opt_suffix);
  if (suffix_string.length() > MAX_SUFFIX_LEN) {
    error << "Maximum number of characters allowed as the value for "
          << "--suffix are " << MAX_SUFFIX_LEN << endl;
    ret_val = 1;
    goto end;
  }

  if (check_suffix()) {
    error << "Invalid string for --suffix option. Either use default value for "
          << "the option or provide a string with alphanumeric characters "
          << "and/or _ only." << endl;
    ret_val = 1;
    goto end;
  }

  if ((ret_val = execute_command(openssl_check,
                                 "Could not find OpenSSL on the system"))) {
    goto end;
  } else {
    char save_wd[FN_REFLEN];
    bool files_exist = false;
    Sql_string_t verify("openssl verify -CAfile ");

    if (my_getwd(save_wd, FN_REFLEN - 1, MYF(MY_WME))) {
      error << "Error saving current working directory" << endl;
      ret_val = 1;
      goto end;
    }

    if (my_setwd(dir_string.to_str().c_str(), MYF(MY_WME))) {
      error << "Error changing working directory" << endl;
      ret_val = 1;
      goto end;
    }
#if HAVE_CHOWN
    if (opt_userid && geteuid() == 0) {
      user_info = getpwnam(opt_userid);
      if (!user_info) {
        error << "Error fetching user information" << endl;
        ret_val = 1;
        goto end;
      }
    }
#endif /* HAVE_CHOWN */

    /*
      SSL Certificate Generation.
      1. Check for ca.pem, server_cert.pem and server_key.pem at
         the directory location provided by --dir.
      2. If none of these files are present, generate following
         files:
         ca.pem, ca_key.pem
         server_cert.pem, server_key.pem
         client_cert.pem, client_key.pem
      3. If everything goes smoothly, set permission on files.
    */

    files_exist = file_exists(cert_files[CA_CERT]) ||
                  file_exists(cert_files[SERVER_CERT]) ||
                  file_exists(cert_files[SERVER_KEY]);

    if (files_exist) {
      info << "Certificate files are present in given dir. Skipping generation."
           << endl;
    } else {
      Sql_string_t empty_string("");
      X509_key x509_key(suffix_string);
      X509_cert x509_cert;
      X509v3_ext_writer x509v3_ext_writer;

      /* Delete existing files if any */
      remove_file(cert_files[CA_REQ], false);
      remove_file(cert_files[SERVER_REQ], false);
      remove_file(cert_files[CLIENT_REQ], false);
      remove_file(cert_files[CLIENT_CERT], false);
      remove_file(cert_files[CLIENT_KEY], false);
      remove_file(cert_files[OPENSSL_RND], false);

      /* Remove existing v3 extension files */
      remove_file(ext_files[CAV3_EXT], false);
      remove_file(ext_files[CERTV3_EXT], false);

      /* Create v3 extension files */
      if (x509v3_ext_writer(ext_files[CAV3_EXT], ext_files[CERTV3_EXT]))
        goto end;

      /* Generate CA Key and Certificate */
      if ((ret_val =
               execute_command(x509_key("_Auto_Generated_CA_Certificate",
                                        cert_files[CA_KEY], cert_files[CA_REQ]),
                               "Error generating ca_key.pem and ca_req.pem")))
        goto end;

      if ((ret_val = execute_command(
               x509_cert(cert_files[CA_REQ], cert_files[CA_CERT], 1, true,
                         cert_files[CA_KEY], empty_string, ext_files[CAV3_EXT]),
               "Error generating ca_cert.pem")))
        goto end;

      /* Generate Server Key and Certificate */
      if ((ret_val = execute_command(
               x509_key("_Auto_Generated_Server_Certificate",
                        cert_files[SERVER_KEY], cert_files[SERVER_REQ]),
               "Error generating server_key.pem and server_req.pem")))
        goto end;

      if ((ret_val = execute_command(
               x509_cert(cert_files[SERVER_REQ], cert_files[SERVER_CERT], 2,
                         false, cert_files[CA_KEY], cert_files[CA_CERT],
                         ext_files[CERTV3_EXT]),
               "Error generating server_cert.pem")))
        goto end;

      /* Generate Client Key and Certificate */
      if ((ret_val = execute_command(
               x509_key("_Auto_Generated_Client_Certificate",
                        cert_files[CLIENT_KEY], cert_files[CLIENT_REQ]),
               "Error generating client_key.pem and client_req.pem")))
        goto end;

      if ((ret_val = execute_command(
               x509_cert(cert_files[CLIENT_REQ], cert_files[CLIENT_CERT], 3,
                         false, cert_files[CA_KEY], cert_files[CA_CERT],
                         ext_files[CERTV3_EXT]),
               "Error generating client_cert.pem")))
        goto end;

      /* Verify generated certificates */
      verify.append(cert_files[CA_CERT]);
      verify.append(" ");
      verify.append(cert_files[SERVER_CERT]);
      verify.append(" ");
      verify.append(cert_files[CLIENT_CERT]);
      if ((ret_val = execute_command(
               verify, "Verification of X509 certificates failed.")))
        goto end;

      /* Set permission */
      if ((ret_val = (set_file_pair_permission(cert_files[CA_KEY],
                                               cert_files[CA_CERT]) |
                      set_file_pair_permission(cert_files[SERVER_KEY],
                                               cert_files[SERVER_CERT]) |
                      set_file_pair_permission(cert_files[CLIENT_KEY],
                                               cert_files[CLIENT_CERT]))))
        goto end;

      /* Remove request files : Flag an error if we can't delete them. */
      if ((ret_val = remove_file(cert_files[CA_REQ]))) goto end;

      if ((ret_val = remove_file(cert_files[SERVER_REQ]))) goto end;

      if ((ret_val = remove_file(cert_files[CLIENT_REQ]))) goto end;

      remove_file(cert_files[OPENSSL_RND], false);

      /* Remove existing v3 extension files */
      remove_file(ext_files[CAV3_EXT], false);
      remove_file(ext_files[CERTV3_EXT], false);
    }

    /*
      RSA Key pair generation.
      1. Check if private_key.pem or public_key.pem are present at
         the directory location provided by --dir.
      2. If not, generate private_key.pem, public_key.pem and
         set permission after successful generation.
    */

    files_exist = file_exists(cert_files[PRIVATE_KEY]) ||
                  file_exists(cert_files[PUBLIC_KEY]);

    if (files_exist) {
      info << "RSA key files are present in given dir. Skipping generation."
           << endl;
    } else {
      RSA_priv rsa_priv;
      RSA_pub rsa_pub;

      /* Remove existing file if any */
      remove_file(cert_files[OPENSSL_RND], false);

      if ((ret_val = execute_command(rsa_priv(cert_files[PRIVATE_KEY]),
                                     "Error generating private_key.pem")))
        goto end;

      if ((ret_val = execute_command(
               rsa_pub(cert_files[PRIVATE_KEY], cert_files[PUBLIC_KEY]),
               "Error generating public_key.pem")))
        goto end;
      /* Set Permission */
      if ((ret_val = set_file_pair_permission(cert_files[PRIVATE_KEY],
                                              cert_files[PUBLIC_KEY])))
        goto end;

      remove_file(cert_files[OPENSSL_RND], false);
    }

    if (my_setwd(save_wd, MYF(MY_WME))) {
      error << "Error changing working directory" << endl;
      ret_val = 1;
      goto end;
    }
  }

  info << "Success!" << endl;

end:

  umask(saved_umask);
  free_resources();

  return ret_val;
}
